### PR TITLE
ThirdPartyResource example: added watcher example, code cleanup

### PR DESCRIPTION
**Pull request by nilebox**
**NOTE**: This is a cleaned and updated version of PR https://github.com/kubernetes/kubernetes/pull/43027

**What this PR does / why we need it**:
An example of using go-client for watching on ThirdPartyResource events (create/update/delete).

### CODE REVIEW

**Comment by nilebox** [comment]
@sttts @caesarxuchao Please see this updated PR.
I have removed the unneccessary and contradictory code which has blocked the previous PR from being merged.
As there is a duplicate https://github.com/kubernetes/kubernetes/pull/45263, it proves the importance of such an example.

File: staging/src/k8s.io/client-go/examples/third-party-resources/main.go [hunk 1]
```diff
@@ -52,24 +50,29 @@ func main() {
...
 
 			result, err := clientset.ExtensionsV1beta1().ThirdPartyResources().Create(tpr)
 			if err != nil {
 				panic(err)
 			}
 			fmt.Printf("CREATED: %#v\nFROM: %#v\n", result, tpr)
+
+			// See the issue https://github.com/kubernetes/features/issues/95
+			// ("Make new TPRs available immediately after the create request succeeds")
+			fmt.Print("Sleeping to make sure the TPR is processed and available")
```
Comment (MikeSpreitzer) [inline staging/src/k8s.io/client-go/examples/third-party-resources/main.go]: How about upgrading the comment, if not the printed message, to point to the existing issue (https://github.com/kubernetes/features/issues/95) that notes the need for a delay (among many other things, sigh)?

**Review by MikeSpreitzer — COMMENTED** [review]
If we are trying to give a good example, that would use glog instead of fmt, right?
But that does not need to hold up this PR, it could be done in a later PR.

File: staging/src/k8s.io/client-go/examples/third-party-resources/main.go (hunk 1 above)
Comment (nilebox) [inline staging/src/k8s.io/client-go/examples/third-party-resources/main.go]: @MikeSpreitzer It does have a comment above pointing to this issue already :)
Not sure if adding the issue link to the printed message helps...

**Comment by sttts** [comment]
@k8s-bot ok to test

**Comment by sttts** [comment]
@nilebox will take a look. In general, try to avoid creating new PRs as this remove the context of previous discussions.

**Comment by sttts** [comment]
@nilebox am still not really happy with the example. IMO we have to simplify it quite a lot. Let me contribute an integration test on-top of your code. This is far too complex to be without test.

**Comment by nilebox** [comment]
@sttts sure, integration test would be helpful.

File: staging/src/k8s.io/client-go/examples/third-party-resources/controller.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+		example := obj.(*tprv1.Example)
+		fmt.Printf("Existing example: %#v\n", example)
+	}
+
+	go controller.Run(ctx.Done())
+
+	return controller, nil
+}
+
+func onAdd(obj interface{}) {
```
Comment (sttts) [inline staging/src/k8s.io/client-go/examples/third-party-resources/controller.go]: usually these are methods of the controller.

File: staging/src/k8s.io/client-go/examples/third-party-resources/main.go [hunk 3]
```diff
@@ -52,66 +53,65 @@ func main() {
...
 
-	err = tprclient.Get().
-		Resource("examples").
-		Namespace(v1.NamespaceDefault).
+	// The sleep below is just to make sure that the watcher.Run() goroutine has successfully executed
+	// and the watcher is handling the events about Example TPR instances.
+	// In the normal application there is no need for it, because:
+	// 1. It's unlikely to create a watcher and a TPR instance at the same time in the same application.
+	// 2. The application with watcher would most probably keep running instead of exiting right after the watcher startup.
+	time.Sleep(5 * time.Second)
```
Comment (sttts) [inline staging/src/k8s.io/client-go/examples/third-party-resources/main.go]: Do we pass a zero resourceVersion to the Watch request? We should get all old events that are in the apiserver watch cache.

File: staging/src/k8s.io/client-go/examples/third-party-resources/main.go [hunk 4]
```diff
@@ -52,66 +53,65 @@ func main() {
...
+
+	err = exampleClient.Get().
+		Resource(tprv1.ExampleResourcePlural).
+		Namespace(apiv1.NamespaceDefault).
 		Name("example1").
 		Do().Into(&example)
 
 	if err != nil {
-		if errors.IsNotFound(err) {
+		if apierrors.IsNotFound(err) {
```
Comment (sttts) [inline staging/src/k8s.io/client-go/examples/third-party-resources/main.go]: Just try to post the tpr and check for errors.IsAlreadyExists. No need for the whole Get+Post logic here.

Comment (sttts) [inline staging/src/k8s.io/client-go/examples/third-party-resources/main.go]: Compare the tpr creation itself. We do the same there.

File: test/integration/thirdparty/client-go_test.go [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+
+	t.Logf("Waiting for TPR %q to show up", exampletprv1.ExampleResourcePlural)
+	if err := exampleclient.WaitForExampleResource(exampleClient); err != nil {
+		t.Fatalf("ThirdPartyResource examples did not show up: %v", err)
+	}
+	t.Logf("TPR %q is active", exampletprv1.ExampleResourcePlural)
+
```
Comment (sttts) [inline test/integration/thirdparty/client-go_test.go]: Here we need the controller as well. For that you have to move the ExampleController into a `controller` sub-package.

File: staging/src/k8s.io/client-go/examples/third-party-resources/main.go (hunk 3 above)
Comment (nilebox) [inline staging/src/k8s.io/client-go/examples/third-party-resources/main.go]: Yes, I think it's fine to get all the old events?
This sleep is only to ensure that we handle the example created below (on the first run, when we don't have any old objects and will create a new one).

File: test/integration/thirdparty/client-go_test.go [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+		ExampleClient: exampleClient,
+	}
+
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	defer cancelFunc()
+	err = controller.Run(ctx)
+	if err != nil {
+		t.Fatalf("unexpected error starting a TPR controller: %v", err)
+	}
+
```
Comment (caesarxuchao) [inline test/integration/thirdparty/client-go_test.go]: nit: remove the empty lines.

**Comment by caesarxuchao** [comment]
Thanks @nilebox! @sttts could you apply the lgtm when you approve it? It generally looks good to me. I don't have time to review closely. Thanks.

File: staging/src/k8s.io/client-go/examples/third-party-resources/client/client.go [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/client-go/rest"
+
+	tprv1 "k8s.io/client-go/examples/third-party-resources/apis/tpr/v1"
+)
+
+func NewClient(cfg *rest.Config) (*rest.RESTClient, *runtime.Scheme, error) {
+	scheme := runtime.NewScheme()
+	if err := tprv1.AddToScheme(scheme); err != nil {
```
Comment (caesarxuchao) [inline staging/src/k8s.io/client-go/examples/third-party-resources/client/client.go]: Is List/Watch working for tpr client? I guess they are since the controller needs them but i don't understand why they are working. List/Watch are using the global scheme/codec, see this [line](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/kubernetes/typed/core/v1/pod.go#L156), not the scheme constructed here. I thought the global scheme/codec would complain not recognizing the tpr groupVersionKind.

Comment (caesarxuchao) [inline staging/src/k8s.io/client-go/examples/third-party-resources/client/client.go]: Nevermind, the apimachinery code is very tolerant for unversioned Kind (ListOptions is registered as one) and thus there is no error.

File: staging/src/k8s.io/client-go/examples/third-party-resources/controller/controller.go [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+
+		// Your custom resource event handlers.
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.onAdd,
+			UpdateFunc: c.onUpdate,
+			DeleteFunc: c.onDelete,
+		})
+
+	// store can be used to List and Get
+	for _, obj := range store.List() {
```
Comment (ash2k) [inline staging/src/k8s.io/client-go/examples/third-party-resources/controller/controller.go]: `store.List()` will always return an empty slice here because Informer has not been started yet and hadn't done any list/watch operations yet. Maybe add `cache.WaitForCacheSync(ctx.Done(), controller.HasSynced)` after the `controller.Run()` to make sure it has synced and move this `for` after it?

File: staging/src/k8s.io/client-go/examples/third-party-resources/controller/controller.go [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+
+	<-ctx.Done()
+	return ctx.Err()
+}
+
+func (c *ExampleController) watchExamples(ctx context.Context) (cache.Controller, error) {
+	source := cache.NewListWatchFromClient(
+		c.ExampleClient,
+		tprv1.ExampleResourcePlural,
+		apiv1.NamespaceAll,
```
Comment (ash2k) [inline staging/src/k8s.io/client-go/examples/third-party-resources/controller/controller.go]: From what I understand from other code, the convention is to use `metav1.Namespace*` constants? `apiv1.Namespace*` constants are only for the package where they reside?

File: staging/src/k8s.io/client-go/examples/third-party-resources/controller/controller.go [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+
+func (c *ExampleController) onUpdate(oldObj, newObj interface{}) {
+	oldExample := oldObj.(*tprv1.Example)
+	newExample := newObj.(*tprv1.Example)
+	fmt.Printf("[WATCH] OnUpdate oldObj: %s\n", oldExample.Metadata.SelfLink)
+	fmt.Printf("[WATCH] OnUpdate newObj: %s\n", newExample.Metadata.SelfLink)
+}
+
+func (c *ExampleController) onDelete(obj interface{}) {
+	example := obj.(*tprv1.Example)
```
Comment (ash2k) [inline staging/src/k8s.io/client-go/examples/third-party-resources/controller/controller.go]: It would be awesome for completeness to show that this may fail sometimes - see https://github.com/atlassian/smith/pull/54 for example. I believe that is what all listeners in Kubernetes codebase do.

File: staging/src/k8s.io/client-go/examples/third-party-resources/main.go [hunk 11]
```diff
@@ -52,83 +54,63 @@ func main() {
...
+		},
+		Spec: tprv1.ExampleSpec{
+			Foo: "hello",
+			Bar: true,
+		},
+	}
+	var result tprv1.Example
+	err = exampleClient.Post().
+		Resource(tprv1.ExampleResourcePlural).
+		Namespace(apiv1.NamespaceDefault).
```
Comment (ash2k) [inline staging/src/k8s.io/client-go/examples/third-party-resources/main.go]: `metav1.NamespaceDefault`?

File: test/integration/thirdparty/client-go_test.go [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+	"context"
+)
+
+func TestClientGoThirdPartyResourceExample(t *testing.T) {
+	_, s := framework.RunAMaster(framework.NewIntegrationTestMasterConfig())
+	defer s.Close()
+
+	config := &rest.Config{Host: s.URL, ContentConfig: rest.ContentConfig{NegotiatedSerializer: api.Codecs}}
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
```
Comment (ash2k) [inline test/integration/thirdparty/client-go_test.go]: nit: I think at least some other packages use `assert.NoError(err)`/`require.NoError(err)` to reduce boilerplate.

File: staging/src/k8s.io/client-go/examples/third-party-resources/client/client.go (hunk 7 above)
Comment (nilebox) [inline staging/src/k8s.io/client-go/examples/third-party-resources/client/client.go]: @caesarxuchao yeah, looks like there were some changes recently, it was breaking before but works fine now, at least for the code used in the example.

File: staging/src/k8s.io/client-go/examples/third-party-resources/controller/controller.go (hunk 8 above)
Comment (nilebox) [inline staging/src/k8s.io/client-go/examples/third-party-resources/controller/controller.go]: idk, maybe just remove this code to keep the PR simple...

File: staging/src/k8s.io/client-go/examples/third-party-resources/controller/controller.go (hunk 9 above)
Comment (nilebox) [inline staging/src/k8s.io/client-go/examples/third-party-resources/controller/controller.go]: Not sure, all the examples using `NewListWatchFromClient()` I've found are invoking `apiv1.NamespaceAll`

File: staging/src/k8s.io/client-go/examples/third-party-resources/controller/controller.go (hunk 10 above)
Comment (nilebox) [inline staging/src/k8s.io/client-go/examples/third-party-resources/controller/controller.go]: Good point, but I think it should be a separate PR. This one is already XL...

**Comment by nilebox** [comment]
@sttts I have added a bunch of code for testing that controller actually processes the Example instance by changing its status. Also, it helped me replacing sleeps with polling for expected condition. Added this to an integration test as well, please review.

File: staging/src/k8s.io/client-go/examples/third-party-resources/apis/tpr/v1/types.go [hunk 13]
```diff
@@ -23,25 +23,49 @@ import (
...
 
 type Example struct {
 	metav1.TypeMeta `json:",inline"`
-	Metadata        metav1.ObjectMeta `json:"metadata"`
+	// WARNING: Don't call the field below ObjectMeta to avoid the issue with ugorji
+	// https://github.com/kubernetes/client-go/issues/8#issuecomment-285333502
+	Metadata metav1.ObjectMeta `json:"metadata"`
 
 	Spec ExampleSpec `json:"spec"`
+
```
Comment (sttts) [inline staging/src/k8s.io/client-go/examples/third-party-resources/apis/tpr/v1/types.go]: nit: no empty line

File: staging/src/k8s.io/client-go/examples/third-party-resources/apis/tpr/v1/types.go [hunk 14]
```diff
@@ -23,25 +23,49 @@ import (
...
-	Bar bool   `json:"bar"`
-}
+const ExampleResourcePlural = "examples"
 
 type Example struct {
 	metav1.TypeMeta `json:",inline"`
-	Metadata        metav1.ObjectMeta `json:"metadata"`
+	// WARNING: Don't call the field below ObjectMeta to avoid the issue with ugorji
+	// https://github.com/kubernetes/client-go/issues/8#issuecomment-285333502
+	Metadata metav1.ObjectMeta `json:"metadata"`
```
Comment (sttts) [inline staging/src/k8s.io/client-go/examples/third-party-resources/apis/tpr/v1/types.go]: @nilebox in my last PR I had removed this Metadata and everything kept working. I doubt we need this workaround.

File: staging/src/k8s.io/client-go/examples/third-party-resources/apis/tpr/v1/types.go [hunk 15]
```diff
@@ -23,25 +23,49 @@ import (
...
+	ExampleStateCreated   ExampleState = "Created"
+	ExampleStateProcessed ExampleState = "Processed"
+)
+
 type ExampleList struct {
 	metav1.TypeMeta `json:",inline"`
-	Metadata        metav1.ListMeta `json:"metadata"`
+	// WARNING: Don't call the field below ListMeta to avoid the issue with ugorji
+	// https://github.com/kubernetes/client-go/issues/8#issuecomment-285333502
+	Metadata metav1.ListMeta `json:"metadata"`
```
Comment (sttts) [inline staging/src/k8s.io/client-go/examples/third-party-resources/apis/tpr/v1/types.go]: same here

File: staging/src/k8s.io/client-go/examples/third-party-resources/apis/tpr/v1/types.go (hunk 14 above)
Comment (nilebox) [inline staging/src/k8s.io/client-go/examples/third-party-resources/apis/tpr/v1/types.go]: @sttts completely removed? But having metadata (a field of ObjectMeta types) is useful, why would we remove it? Also, Metadata fields are present in current master (https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/examples/third-party-resources/types.go#L33).

Comment (nilebox) [inline staging/src/k8s.io/client-go/examples/third-party-resources/apis/tpr/v1/types.go]: @sttts I think you meant keep it named ObjectMeta/ListMeta. Just checked, and yes, it seems to work now... Also it means there is no need of implementing `GetObjectKind` and other methods. Looks like some recent change has fixed this issue.
Will get rid of it then.

Comment (nilebox) [inline staging/src/k8s.io/client-go/examples/third-party-resources/apis/tpr/v1/types.go]: @sttts looks like we don't need other ugorji workarounds below either. Awesome.

**Comment by nilebox** [comment]
@sttts all the checks succeeded except "Jenkins verification". Not sure why it fails.

File: staging/src/k8s.io/client-go/examples/third-party-resources/apis/tpr/v1/types.go (hunk 14 above)
Comment (sttts) [inline staging/src/k8s.io/client-go/examples/third-party-resources/apis/tpr/v1/types.go]: I meant to embed the ObjectMeta directly, without `Metadata`.

**Comment by nilebox** [comment]
@sttts all checks succeded, need lgtm and squashing commits

**Comment by sttts** [comment]
lgtm. Please squash and let me know when I should label.

**Comment by nilebox** [comment]
@sttts done.

**Comment by sttts** [comment]
Self-approving due to https://github.com/kubernetes/kubernetes/pull/45463#issuecomment-299988184

**Comment by nilebox** [comment]
@k8s-bot unit test this
